packages/client-react-streaming/src/transportedQueryRef.ts

```diff
@@ -29,92 +55,166 @@ export type TransportedQueryRefOptions = TransportedOptions &
+      __eventStream
+    );
+  };
 }
 
-export function createTransportedQueryRef<TData, TVariables>(
-  options: TransportedQueryRefOptions,
+export function createTransportedQueryRef<
+  TData,
+  TVariables extends OperationVariables,
+>(
+  query: DocumentNode | TypedDocumentNode<TData, TVariables>,
+  options: PreloadTransportedQueryOptions<NoInfer<TVariables>, TData>,
   queryKey: string,
-  _promise: Promise<any>
-): InternalTransportedQueryRef<TData, TVariables> {
-  const ref: InternalTransportedQueryRef<TData, TVariables> = {
-    __transportedQueryRef: true,
-    options,
-    queryKey,
+  stream: ReadableStream<ReadableStreamLinkEvent>
+): TransportedQueryRef<TData, TVariables> {
+  return {
+    $__apollo_queryRef: {
+      options: sanitizeForTransport(serializeOptions({ query, ...options })),
+      queryKey,
+      stream: stream.pipeThrough(new JSONEncodeStream()),
+    },
   };
-  /*
-  Temporarily disabled - see https://github.com/apollographql/apollo-client-nextjs/issues/332
-  This causes a dev-mode warning:
-      Warning: Only plain objects can be passed to Client Components from Server Components. Classes or other objects with methods are not supported.
-      <... queryRef={{__transportedQueryRef: true, options: ..., queryKey: ...}}>
-                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-  */
-  // Object.defineProperty(ref, "toPromise", {
-  //   value: () => promise.then(() => ref),
-  //   enumerable: false,
-  // });
-  return ref;
 }
 
 export function reviveTransportedQueryRef(
-  queryRef: InternalTransportedQueryRef,
+  queryRef: TransportedQueryRef,
   client: ApolloClient<any>
-): [QueryRef<any, any>, CacheKey] {
-  const hydratedOptions = deserializeOptions(queryRef.options);
+): asserts queryRef is TransportedQueryRef &
+  ReturnType<typeof wrapQueryRef<any, any>> & { _hydrated: CacheKey } {
+  const {
+    $__apollo_queryRef: { options, stream, queryKey },
+  } = queryRef;
+  const hydratedOptions = deserializeOptions(options);
   const cacheKey: CacheKey = [
     hydratedOptions.query,
     canonicalStringify(hydratedOptions.variables),
-    queryRef.queryKey,
+    queryKey,
   ];
-  if (queryRef.__transportedQueryRef === true) {
-    queryRef.__transportedQueryRef = wrapQueryRef(
-      getSuspenseCache(client).getQueryRef(cacheKey, () =>
-        client.watchQuery(hydratedOptions)
-      )
+  if (!queryRef._hydrated) {
+    queryRef._hydrated = cacheKey;
+    const internalQueryRef = getSuspenseCache(client).getQueryRef(
+      cacheKey,
+      () =>
+        client.watchQuery({
+          ...hydratedOptions,
+          fetchPolicy: "network-only",
+          context: skipDataTransport(
+            readFromReadableStream(stream.pipeThrough(new JSONDecodeStream()), {
+              ...hydratedOptions.context,
+              queryDeduplication: true,
+            })
+          ),
+        })
     );
+    Object.assign(queryRef, wrapQueryRef(internalQueryRef));
   }
```

Comment:
This is the core "recreate a normal queryRef from a transported queryRef" implementation